File: orario.cpp
#include "orario.h"

//costruttore
orario::orario(int o, int m, int s) {
    if(o<0 || o>23 || m<0 || m>59 || s<0 || s>59)
        sec=0;
    else
       sec=(o*60*60)+(m*60)+s;
}

//costruttore di copia deep
orario::orario(const orario& o) : sec(o.getSec()) {}    //da fare, non lo so fare al momento

//distruttore
orario::~orario() {}

int orario::getOre() const {
    return sec/3600;
}

int orario::getMinuti() const {
    return (sec/60)%60;
}

int orario::getSecondi() const {
    return sec%60;
}

int orario::getSec() const {
    return sec;
}

bool orario::operator==(const orario& o) const {
    return o.sec==sec;
}

orario operator+(const orario& a, const orario& b) {
    int sec=0, min=0, ore=0;

    sec=a.getSecondi()+b.getSecondi();
    min=a.getMinuti()+b.getMinuti()+sec/60;
    sec=sec%60;
    ore=a.getOre()+b.getOre()+min/60;
    min=min%60;
    ore=ore%24;

    orario aux(ore, min, sec);
    return aux;
}

orario operator-(const orario& a, const orario& b) {
    int sec=0, min=0, ore=0;

    if(b.getSecondi()>a.getSecondi()) {
        sec=60+(a.getSecondi()-b.getSecondi());
        min=min+1;
    }
    else
        sec=a.getSecondi()-b.getSecondi();

    if(b.getMinuti()+min>a.getMinuti()) {
        min=60+(a.getMinuti()-(b.getMinuti()+min));
        ore=ore+1;
    }
    else
        min=a.getMinuti()-(b.getMinuti()+min);

    if(b.getOre()+ore>a.getOre())
        ore=24+(a.getOre()-(b.getOre()+ore));
    else
        ore=a.getOre()-(b.getOre()+ore);
    orario aux(ore, min, sec);
    return aux;
}


ostream& operator<<(ostream& os, const orario& o) {
    return os << o.getOre() << ":" << o.getMinuti() << ":" << o.getSecondi() << endl;
}

File: progetto/java/projectX/orario.java
package projectX;

import projectX.overflow_error;


public class orario {
	int sec;
	
	public orario() {
		sec=0;
	}
	
	//costruttore
	public orario(int o, int m, int s) throws overflow_error {
	    if(o<0 || o>23 || m<0 || m>59 || s<0 || s>59)
	        throw new overflow_error();
	    else
	        sec=(o*3600)+(m*60)+s;
	};

	//costruttore di copia deep?

	public int getOre() {
	    return sec/3600;
	}

	public int getMinuti() {
	    return (sec/60)%60;
	}

	public int getSecondi() {
	    return sec%60;
	}

	public int getSec() {
	    return sec;
	}

	public void cambiaSec(int s) {
	    sec=s;
	}

	public orario somma(orario o) throws overflow_error {
	    int secT=0, minT=0, oreT=0;

	    secT=o.getSecondi()+sec;
	    minT=o.getMinuti()+secT/60;
	    secT=secT%60;
	    oreT=o.getOre()+minT/60;
	    minT=minT%60;
	    oreT=oreT%24;

	    orario aux=new orario(oreT, minT, secT);
	    return aux;
	}

	public orario sottrazione(orario o) throws overflow_error {
	    int secT=0, minT=0, oreT=0;

	    if(o.getSecondi()>getSecondi()) {
	        secT=60+(getSecondi()-o.getSecondi());
	        //secT=60+(sec%60-o.getSecondi());
	        minT=minT+1;
	    }
	    else
	        secT=getSecondi()-o.getSecondi();

	    if(o.getMinuti()+minT>getMinuti()) {
	        minT=60+(getMinuti()-(o.getMinuti()+minT));
	        oreT=oreT+1;
	    }
	    else
	        minT=getMinuti()-(o.getMinuti()+minT);

	    if(o.getOre()+oreT>getOre())
	        oreT=24+(getOre()-(o.getOre()+oreT));
	    else
	        oreT=getOre()-(o.getOre()+oreT);

	    orario aux=new orario(oreT, minT, secT);
	    return aux;
	}

	public void aggiungiSecondi(int s) {
	    if(s<0)
	        sottraiSecondi(-s);
	    else if(s>0 && s+sec>86400)
	        sec=sec+s-86400*((s+sec)/86400);
	    else if(s>=0)
	        sec=sec+s;
	}

	public void aggiungiMinuti(int m) {
	    if(m<0)
	        sottraiMinuti(-m);
	    else if(m>0 && m*60+sec>86400)
	        sec=sec+(m*60)-86400*((sec+m*60)/86400);
	    else if(m>0)
	        sec=sec+m*60;
	}

	public void aggiungiOre(int o) {
	    if(o<0)
	        sottraiOre(-o);
	    else if(o>0 && o*3600+sec>86400)
	        sec=sec+(o*3600)-86400*((sec+o*3600)/86400);
	    else if(o>0)
	        sec=sec+o*3600;
	}

	public void sottraiSecondi(int s) {
	    if(s<0)
	        aggiungiSecondi(-s);
	    else if(s>0 && sec<s) //giorni precedenti
	        sec=86400-((86400*(s/86400))-(sec-s))%86400;
	    else if(s>=0)
	        sec=sec-s;
	}

	public void sottraiMinuti(int m) {
	    if(m<0)
	        aggiungiMinuti(-m);
	    else if(m>0 && sec<m*60)
	        sec=86400-((86400*((m*60)/86400))-(sec-m*60))%86400;
	    else if(m>0)
	        sec=sec-m*60;
	}

	public void sottraiOre(int o) {
	    if(o<0)
	        aggiungiOre(-o);
	    else if(o>0 && sec<o*3600)
	        sec=86400-((86400*((o*3600)/86400))-(sec-o*3600))%86400;
	    else if(o>0)
	        sec=sec-o*3600;
	}

	//ritorna la velocita' media fatta in un intervallo o per fare km kilometri
	public double distanzaVel(double km) throws overflow_error {
		if(km<=0)
			throw new overflow_error();
		else
			return (km/sec)*3600;
	}


	public orario quantoMancaMezzogiorno() {
	    orario aux=new orario();
	    if(sec > 43200)  //se orario>12:00:00
	        aux.sec=86400-sec+43200;
	    else
	        aux.sec=43200-sec;
	    return aux;
	}

	public orario divisione(int d) {
	    //divido sec in d parti
	    orario aux=new orario();
	    if(sec<d)
	        return aux;
	    else
	        aux.sec=sec/d;
	    return aux;
	}

	public orario formato12h() {
	    orario aux=new orario();
	    if(sec>=43200)
	        aux.sec=sec-43200;
	    else
	        aux.sec=sec;
	    return aux;
	}

	public void clearMem() throws overflow_error {
	    cambiaSec(0);
	}

	//overloading con public void println()
	public void println() {
		System.out.print(this.getOre() + ":" + this.getMinuti() + ":" + this.getSecondi());
	}

}
